[move] introduce bytecode specification

vgao1996 opened the pull request:
## Description
This introduces a procedural macro `bytecode_spec` that allows us to add inline specifications to the Move bytecode instructions defined in Rust:
```rust
#[bytecode_spec]
pub enum Bytecode {
    #[group = "stack_and_local"]
    #[description = "Pop and discard the value at the top of the stack. The value on the stack must be an copyable type."]
    #[semantics = "stack >> _"]
    #[paranoid_post = r#"
        ty_stack >> ty
        assert ty has drop
    "#]
    Pop,

    ..
}
```
The macro will generate a `pub fn spec() -> Vec<BTreeMap<String, String>>`, which can be used to extract the specification defined using the the attributes. 
```json
[
    {
        "description": "Pop and discard the value at the top of the stack. The value on the stack must be an copyable type.",
        "group": "stack_and_local",
        "name": "pop",
        "paranoid_post": "ty_stack >> ty\nassert ty has drop",
        "semantics": "stack >> _",
    },
]
```
The extracted specification can then be used to generate other human readable docs (e.g. pdf, html, LaTeX etc.) The generator(s) will come in future PRs.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [x] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Unit tests

## Key Areas to Review
- `move-binary-format/src/file_format.rs`: full specification for all current bytecode instructions
- `move-bytecode-spec`: procedural macro to allow the specification to be extracted

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

vgao1996 commented on third_party/move/move-binary-format/src/file_format.rs:
````diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
-    ///
-    /// ```..., arg_val(1), ..., arg_val(n) -> ..., return_val(1), ..., return_val(n)```
+
+    // TODO: double check semantics
````
Note: it's been some time since I wrote the semantics here so I'd like to double check it. Will get to it a bit later but will definitely get it done before landing.

ziaptos commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+    #[group = "stack_and_local"]
+    #[description = "Pop and discard the value at the top of the stack. The value on the stack must be an copyable type."]
+    #[semantics = "stack >> _"]
+    #[paranoid_post = r#"
```
I'd prefer not calling this `paranoid`. Maybe `runtime_type_check_epilogue` and `runtime_type_check_prologue`. Calling it just `paranoid` (not even `paranoid_type_check`).

ziaptos reviewed (changes requested):
Looks straight-forward and good. But I'm concern about the `paranoid` field - see inline comment.

Also, should we stop using `move-X-Y-Z` directory names, as they are all obviously inside the `move` directory?

georgemitenkov commented on third_party/move/move-bytecode-spec/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+name = "move-bytecode-spec"
+version = "0.1.0"
+authors = ["Aptos Labs"]
+description = "Move compiler based on stackless bytecode"
```
Description is wrong :)

georgemitenkov commented on third_party/move/move-bytecode-spec/tests/tests.rs:
```diff
@@ -0,0 +1,140 @@
+}
+
+#[test]
+#[allow(dead_code)]
```
Why these are dead? These are tests?

vgao1996 replied:
This is to suppress the `never-constructed enum variants` warning. The compiler cannot see that the variants are being processed by the procedural macro.

georgemitenkov commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+    "#]
+    #[semantics = r#"
+        call_stack >> current_frame 
+        // The frame of the function being return from is dropped.
```
typo: returned? Let's proof-read the comments before landing

georgemitenkov commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
 }
 
+#[test]
+fn tttt() {
```
Maybe a nicer name? :)

vgao1996 replied:
My bad, used this as an ad-hoc entry point to print out the extracted spec. Will get it removed.

georgemitenkov reviewed (approved):
Please address Zi's comment before landing

georgemitenkov commented:
For directories, good question. We also have everything called `aptos-...`. Would be good to make this consistent (not prefixed at all?)

vgao1996 commented:
@georgemitenkov @ziaptos Yeah the discussion on directory naming is great. Although I feel like we should handle it separately for the sake of consistency. Need to first reach agreement on a convention, and then perform batch renaming.

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+
+    #[group = "bitwise"]
+    #[description = r#"
+        Shift the (second top value) right (top value) bits and pushes the result on the stack.
```
Are there any restrictions on the shifted value being `u64` (document it eitherway)?

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+            stack << (lhs __shift_left__ rhs)
+    "#]
+    #[paranoid_post = r#"
+        ty_stack >> left_ty 
```
These do not seem to reflect what is being described in the semantics description.

vgao1996 replied:
Fixed. The original paranoid mode code used the wrong notations.

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+            stack << (lhs __shift_right__ rhs)
+    "#]
+    #[paranoid_post = r#"
+        ty_stack >> left_ty 
```
Again, incorrect reflection of the semantics description.

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+        stack << vector[elem_1, .., elem_n]
+    "#]
+    #[paranoid_post = r#"
+        elem_ty << instantiate elem_ty 
```
```suggestion
        elem_ty = instantiate elem_ty 
```

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+        Abort the execution if there are not enough number of elements on the stack 
+        to pack from or they do not have the same type identified by the `elem_ty_idx`.
+    "#]
+    #[static_operands = "[elem_ty_idx] [n]"]
```
In some places, we use `[u64_value]`, but here we use `[n]`?

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+    #[paranoid_post = r#"
+        elem_ty << instantiate elem_ty 
+        ty_stack >> ty 
+        assert ty == &elem_ty 
```
From the semantics description, it seems like this should be a reference to a vector type, not reference to an element of vector?

vgao1996 replied:
Yes, I'm just using the notion `&elem_ty` to indicate that this is a reference type created by prefixing `elem_ty` with `&`. 

Might be a bit confusing if you think of it in terms rust, but within the context of paranoid mode we only have types and no values so I guess it should be fine. Later we can take another look at the pseudocode syntax and see if we can improve it.

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+        ty_stack >> idx_ty 
+        assert idx_ty == u64 
+        ty_stack >> ref_ty 
+        assert ref_ty == &vector<elem_ty> 
```
Should the reference to the vector be `&mut` instead?

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+    #[group = "vector"]
+    #[description = r#"
+        Swaps the elements at two indices in the vector. 
+        Abort the execution if any of the indice is out of bounds.
```
```suggestion
        Abort the execution if any of the indices are out of bounds.
```

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+    #[static_operands = "[u16_value]"]
+    #[semantics = "stack << u16_value"]
+    #[paranoid_post = "ty_stack << u16"]
     LdU16(u16),
```
Weird that these show up down here instead of with the other `LdUx` above. Would be nice if they were together.

vgao1996 replied:
The reason why these instructions appear at the end is because they were added at a later time. 

Now since we don't derive the serialization of the instructions it may very well be fine to change the order, but that's better done in a separate PR, with adequate verification.

vineethk reviewed (commented):
I really like this effort, and the PR looks generally pretty good to me.

I have a bunch of comments, will approve after they are addressed.

vgao1996 commented on third_party/move/move-bytecode-spec/src/lib.rs:
```diff
@@ -0,0 +1,191 @@
+            if let Ok(Meta::NameValue(nv)) = attr.parse_meta() {
+                if let Some(attr_name) = nv.path.get_ident() {
+                    let attr_name = attr_name.to_string();
+                    if KNOWN_ATTRIBUTES.contains_key(attr_name.as_str()) {
```
Not sure if we should do that as unknown attributes could belong to other macros. Stripping the ones you know should put us in a safer position.

vineethk replied:
Ah, fair point. I was just worried about potential typos in the attribute names. Perhaps eventually, we can make some subset of these required attributes, and check that each variant has those.

vgao1996 commented on third_party/move/move-bytecode-spec/src/lib.rs:
```diff
@@ -0,0 +1,191 @@
+        match map_entries.entry("name".to_string()) {
+            btree_map::Entry::Occupied(_entry) => (),
+            btree_map::Entry::Vacant(entry) => {
+                entry.insert(upper_camel_to_lower_snake_case(&variant_name));
```
For now just personal preference. I find "ld_u8" a bit more readable than "LdU8", as the latter is very cramped, making it hard for my brain to tokenize it.

Totally open for discussion though.

vineethk replied:
I don't have a strong opinion either way, but having it be `LdU8` means I can copy it from code and search verbatim in whatever semantics document we produce (eg, in the html page) for it.

vineethk replied:
I see that you used the lower_snake_case convention in the gas profiler output as well. So keeping this consistent is better - in light of this, I would suggest keeping this as is.

vgao1996 commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+        stack << field_n
+    "#]
+    #[paranoid_post = r#"
+        ty_stack >> ty 
```
Checked by the bytecode verifier.

vgao1996 commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1340,501 +1341,1256 @@ pub struct CodeUnit {
+        
+        if func.is_native()
+            call_native(func.name, ty_args, args = [arg_1, .., arg_n])
+            current_frame.pc += 1
```
In my original doc I have a line stating that `current_frame.pc += 1` is implicit for all non-control-flow instructions.

For the control-flow instructions however, changes to PC must be explicitly specified due to them being non-trivial.

vgao1996 commented:
@vineethk @georgemitenkov @ziaptos I've addressed all your comments except for the derivation of static operands -- this one requires a bit more hacking with parsing so I'll probably come back to it later.

vgao1996 commented:
@vineethk For static operands, I think it would be ideal to have a sub attribute `#[operand_name]` attached to each operand. 
```
LdU32(#[operand_name = "value_to_load"] u32)
```
This will be translated to `[value_to_load: u32]`

vineethk commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1335,503 +1336,1277 @@ pub struct CodeUnit {
+        The operands can be of any (but the same) primitive integer type.
+    "#]
+    #[semantics = r#"
+        stack >> (rhs: int_ty)
```
In the semantics, we pop off rhs, then lhs. In the runtime_check_epilogue, we pop off left_ty and then right_ty. This happens for several bytecodes and should be fixed.

vineethk replied:
@vgao1996 Just wanted to check if this was fixed.

vgao1996 replied:
Not yet, but will fix this + check similar instructions before landing.

vineethk reviewed (approved):
I have one other comment added (about the ordering of lhs and rhs in semantics vs. paranoid mode). Other than that, LGTM.

ziaptos reviewed (commented):
LGTM

ziaptos reviewed (approved):
LGTM

vgao1996 reviewed (commented):
@vineethk I've fixed the lhs/rhs confusion, which originated from the paranoid mode code itself, which I also fixed. Now trying to land the PR
